File: core/src/main/java/nl/t64/game/rpg/components/party/EquipContainer.java
package nl.t64.game.rpg.components.party;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;


class EquipContainer {

    private static final int NUMBER_OF_EQUIPMENT_SLOTS = 14;

    private final Map<String, InventoryItem> equipment;

    private EquipContainer() {
        this.equipment = new HashMap<>(NUMBER_OF_EQUIPMENT_SLOTS);
        this.equipment.put(InventoryGroup.WEAPON.name(), null);
        this.equipment.put(InventoryGroup.SHIELD.name(), null);
        this.equipment.put(InventoryGroup.ACCESSORY.name(), null);
        this.equipment.put(InventoryGroup.HELMET.name(), null);
        this.equipment.put(InventoryGroup.NECKLACE.name(), null);
        this.equipment.put(InventoryGroup.SHOULDERS.name(), null);
        this.equipment.put(InventoryGroup.CHEST.name(), null);
        this.equipment.put(InventoryGroup.CLOAK.name(), null);
        this.equipment.put(InventoryGroup.BRACERS.name(), null);
        this.equipment.put(InventoryGroup.GLOVES.name(), null);
        this.equipment.put(InventoryGroup.RING.name(), null);
        this.equipment.put(InventoryGroup.BELT.name(), null);
        this.equipment.put(InventoryGroup.PANTS.name(), null);
        this.equipment.put(InventoryGroup.BOOTS.name(), null);
    }

    @JsonCreator
    private EquipContainer(@JsonProperty("weapon") String weaponId,
                           @JsonProperty("shield") String shieldId,
                           @JsonProperty("chest") String chestId) {
        this();
        var database = InventoryDatabase.getInstance();
        this.equipment.put(InventoryGroup.WEAPON.name(), database.getInventoryItem(weaponId));
        if (shieldId != null) {
            this.equipment.put(InventoryGroup.SHIELD.name(), database.getInventoryItem(shieldId));
        }
        this.equipment.put(InventoryGroup.CHEST.name(), database.getInventoryItem(chestId));
    }

    Optional<InventoryItem> getInventoryItem(InventoryGroup inventoryGroup) {
        return Optional.ofNullable(equipment.get(inventoryGroup.name()));
    }

    void forceSetInventoryItem(InventoryGroup inventoryGroup, InventoryItem inventoryItem) {
        equipment.replace(inventoryGroup.name(), inventoryItem);
    }

    int getStatValueOf(InventoryGroup inventoryGroup, StatItemId statItemId) {
        return getInventoryItem(inventoryGroup).map(inventoryItem -> inventoryItem.getAttributeOfStatItemId(statItemId))
                                               .orElse(0);
    }

    int getSkillValueOf(InventoryGroup inventoryGroup, SkillItemId skillItemId) {
        return getInventoryItem(inventoryGroup).map(inventoryItem -> inventoryItem.getAttributeOfSkillItemId(skillItemId))
                                               .orElse(0);
    }

    int getCalcValueOf(InventoryGroup inventoryGroup, CalcAttributeId calcAttributeId) {
        return getInventoryItem(inventoryGroup).map(inventoryItem -> inventoryItem.getAttributeOfCalcAttributeId(calcAttributeId))
                                               .orElse(0);
    }

    int getSumOfStat(StatItemId statItemId) {
        return equipment.values()
                        .stream()
                        .filter(Objects::nonNull)
                        .mapToInt(inventoryItem -> inventoryItem.getAttributeOfStatItemId(statItemId))
                        .sum();
    }

    int getSumOfSkill(SkillItemId skillItemId) {
        return equipment.values()
                        .stream()
                        .filter(Objects::nonNull)
                        .mapToInt(inventoryItem -> inventoryItem.getAttributeOfSkillItemId(skillItemId))
                        .sum();
    }

    int getSumOfCalc(CalcAttributeId calcAttributeId) {
        return equipment.values()
                        .stream()
                        .filter(Objects::nonNull)
                        .mapToInt(inventoryItem -> inventoryItem.getAttributeOfCalcAttributeId(calcAttributeId))
                        .sum();
    }

}

File: core/src/nl/t64/game/rpg/components/party/inventory/EquipContainer.kt
package nl.t64.game.rpg.components.party.inventory

import nl.t64.game.rpg.components.party.CalcAttributeId
import nl.t64.game.rpg.components.party.skills.SkillItemId
import nl.t64.game.rpg.components.party.stats.StatItemId
import java.beans.ConstructorProperties


class EquipContainer() {

    private val equipment: MutableMap<String, InventoryItem?> = mutableMapOf(
        Pair(InventoryGroup.WEAPON.name, null),
        Pair(InventoryGroup.SHIELD.name, null),
        Pair(InventoryGroup.ACCESSORY.name, null),
        Pair(InventoryGroup.HELMET.name, null),
        Pair(InventoryGroup.NECKLACE.name, null),
        Pair(InventoryGroup.SHOULDERS.name, null),
        Pair(InventoryGroup.CHEST.name, null),
        Pair(InventoryGroup.CLOAK.name, null),
        Pair(InventoryGroup.BRACERS.name, null),
        Pair(InventoryGroup.GLOVES.name, null),
        Pair(InventoryGroup.RING.name, null),
        Pair(InventoryGroup.BELT.name, null),
        Pair(InventoryGroup.PANTS.name, null),
        Pair(InventoryGroup.BOOTS.name, null))

    @ConstructorProperties("weapon", "shield", "chest")
    constructor(weaponId: String, shieldId: String?, chestId: String) : this() {
        this.equipment[InventoryGroup.WEAPON.name] = InventoryDatabase.createInventoryItem(weaponId)
        shieldId?.let { this.equipment[InventoryGroup.SHIELD.name] = InventoryDatabase.createInventoryItem(it) }
        this.equipment[InventoryGroup.CHEST.name] = InventoryDatabase.createInventoryItem(chestId)
    }

    fun hasInventoryItem(itemId: String): Boolean {
        return equipment.values.any { it?.id == itemId }
    }

    fun getInventoryItem(inventoryGroup: InventoryGroup): InventoryItem? {
        return equipment[inventoryGroup.name]
    }

    fun clearAll() {
        equipment.keys.forEach { equipment[it] = null }
    }

    fun clearInventoryItem(inventoryGroup: InventoryGroup) {
        equipment[inventoryGroup.name] = null
    }

    fun forceSetInventoryItem(inventoryGroup: InventoryGroup, inventoryItem: InventoryItem) {
        equipment[inventoryGroup.name] = inventoryItem
    }

    fun getStatValueOf(inventoryGroup: InventoryGroup, statItemId: StatItemId): Int {
        return getInventoryItem(inventoryGroup)?.getAttributeOfStatItemId(statItemId) ?: 0
    }

    fun getSkillValueOf(inventoryGroup: InventoryGroup, skillItemId: SkillItemId): Int {
        return getInventoryItem(inventoryGroup)?.getAttributeOfSkillItemId(skillItemId) ?: 0
    }

    fun getCalcValueOf(inventoryGroup: InventoryGroup, calcAttributeId: CalcAttributeId): Int {
        return getInventoryItem(inventoryGroup)?.getAttributeOfCalcAttributeId(calcAttributeId) ?: 0
    }

    fun getItemsWithMinimalOf(statItemId: StatItemId): List<InventoryItem> {
        return equipment.values.filterNotNull().filter { it.getMinimalAttributeOfStatItemId(statItemId) > 0 }
    }

    fun getSumOfStat(statItemId: StatItemId): Int {
        return equipment.values.filterNotNull().sumOf { it.getAttributeOfStatItemId(statItemId) }
    }

    fun getSumOfSkill(skillItemId: SkillItemId): Int {
        return equipment.values.filterNotNull().sumOf { it.getAttributeOfSkillItemId(skillItemId) }
    }

    fun getSumOfCalc(calcAttributeId: CalcAttributeId): Int {
        return equipment.values.filterNotNull().sumOf { it.getAttributeOfCalcAttributeId(calcAttributeId) }
    }

    fun getWeaponSkill(): SkillItemId? {
        return getInventoryItem(InventoryGroup.WEAPON)?.getAttributeOfMinimal(InventoryMinimal.SKILL) as SkillItemId?
    }

    fun getBonusProtectionWhenArmorSetIsComplete(): Int {
        return getInventoryItem(InventoryGroup.HELMET)
            ?.takeIf { doesEquippedArmorAllHaveSamePrefix(getPrefixOfId(it)) }
            ?.getAttributeOfCalcAttributeId(CalcAttributeId.PROTECTION)
            ?: 0
    }

    private fun getPrefixOfId(inventoryItem: InventoryItem): String {
        val splittedId: List<String> = inventoryItem.id.split("_")
        return splittedId[0] + "_" + splittedId[1]
    }

    private fun doesEquippedArmorAllHaveSamePrefix(prefix: String): Boolean {
        return equipment.values
            .filterNotNull()
            .filter { it.group.isPartArmorOfSet() }
            .count { it.id.startsWith(prefix) } == 9
    }

}
